pageserver/src/tenant.rs

```diff
@@ -1063,34 +1341,77 @@ impl Tenant {
                 "Failed to create timeline {}/{} metadata",
-                new_timeline_id, self.tenant_id
+                self.tenant_id, new_timeline_id
             )
         })?;
 
-        let ancestor = new_metadata
-            .ancestor_timeline()
-            .and_then(|ancestor_timeline_id| timelines.get(&ancestor_timeline_id))
-            .cloned();
         let new_timeline = self
-            .initialize_new_timeline(new_timeline_id, new_metadata, ancestor)
+            .create_timeline_data(new_timeline_id, new_metadata, ancestor)
             .with_context(|| {
                 format!(
-                    "Failed to initialize timeline {}/{}",
-                    new_timeline_id, self.tenant_id
+                    "Failed to create timeline data {}/{}",
+                    self.tenant_id, new_timeline_id,
                 )
             })?;
-
-        match timelines.entry(new_timeline_id) {
-            Entry::Occupied(_) => bail!(
-                "Found freshly initialized timeline {} in the tenant map",
-                new_timeline_id
-            ),
-            Entry::Vacant(v) => {
-                v.insert(Arc::clone(&new_timeline));
-            }
+        if init_layers {
+            new_timeline.layers.write().unwrap().next_open_layer_at = Some(new_timeline.initdb_lsn);
         }
 
-        Ok(new_timeline)
+        Ok(UninitializedTimeline {
+            owning_tenant: self,
+            timeline_id: new_timeline_id,
+            raw_timeline: Some((new_timeline, uninit_mark)),
+        })
+    }
+
+    /// Attempts to create an uninit mark file for the timeline initialization.
+    /// Bails, if the timeline is already loaded into the memory (i.e. initialized before), or the uninit mark file already exists.
+    ///
+    /// This way, we need to hold the timelines lock only for small amount of time during the mark check/creation per timeline init.
+    fn create_timeline_uninit_mark(
+        &self,
+        timeline_id: TimelineId,
+        timelines: &MutexGuard<HashMap<TimelineId, Arc<Timeline>>>,
+    ) -> anyhow::Result<TimelineUninitMark> {
+        let tenant_id = self.tenant_id;
+
+        anyhow::ensure!(
+            timelines.get(&timeline_id).is_none(),
+            "Timeline {tenant_id}/{timeline_id} already exists in pageserver's memory"
+        );
+        let uninit_mark_path = self
+            .conf
+            .timeline_uninit_mark_file_path(tenant_id, timeline_id);
+        anyhow::ensure!(
+            !uninit_mark_path.exists(),
+            "Timeline uninit mark file {} already exists, cannot create it again",
+            uninit_mark_path.display()
+        );
+
+        let timeline_path = self.conf.timeline_path(&timeline_id, &tenant_id);
+        anyhow::ensure!(
+            !timeline_path.exists(),
+            "Timeline {} already exists, cannot create its uninit mark file",
+            timeline_path.display()
+        );
+
+        fs::File::create(&uninit_mark_path)
```

Comment:
I was trying to be on the safe here, when one of the pageserver parts exits, leaving the unint mark, but it does not matter with this code, since `fs::File::create(&uninit_mark_path)` will fail if the file exists.
Removed the existence check.

Reply:
Note: `std::fs::File` explicitly documents that it doesn't fail if the file exists, it truncates: https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.create

I think you are assuming it to be `OpenOptions::create_new` but it's not.

Reply:
So, back to your initial comment, implemented. 
I think it's ok generally, since besides `uninit_mark_path`, we could only leave tmp file and tmp initdb dir created.
The latter gets removed if init is attempted again, the former is overwritten, so should be ok.

Reply:
> The latter gets removed if init is attempted again, the former is overwritten, so should be ok.

I am not sure if I can see this right away even from the review, spent some time yesterday looking for it. Resolving as I am agreeing that it most likely doesn't matter with this code.